# **TUGAS PANDAS 2026** 🐼

---
**Nama:**  Hubert Cendana

**NIM:**  245150207111049

Notebook ini berisi penugasan **Pandas** sebagai latihan lanjutan setelah workshop.

### Aturan singkat
- Kerjakan mandiri.
- Gunakan **Pandas** sebagai alat utama.
- Untuk feature engineering sederhana, kamu boleh menggunakan **NumPy** sebagai pendukung.
- Tulis jawaban langsung di cell yang sudah disediakan.


## **📦 Import Library & Load Dataset**

Dataset yang digunakan adalah **`taxis.csv`**.

Sumber dataset:  
`https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv`


In [2]:
import pandas as pd
import numpy as np

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv"
df = pd.read_csv(url)

df.head()


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


## **📌 Soal 1: Kenali Dataset Terlebih Dahulu**

Tampilkan informasi berikut dari dataset `df`:
1. 5 baris teratas
2. Shape dataset
3. Nama semua kolom
4. Informasi detail data dengan `info()`
5. Jumlah missing value per kolom

**Tujuan soal:** membiasakan diri membaca struktur dataset sebelum mulai analisis.


In [19]:
# Tulis jawabanmu di sini
print("1. 5 baris teratas:\n", df.head())
print("\n2. Shape dataset:\n", df.shape)
print("\n3. Nama kolom:\n", df.columns, "\n\n4. Informasi dataset:\n")
print(df.info())
print("\n5. Jumlah missing values per kolom:\n", df.isnull().sum())

1. 5 baris teratas:
                pickup             dropoff  passengers  distance  fare   tip  \
0 2019-03-23 20:21:09 2019-03-23 20:27:24           1      1.60   7.0  2.15   
1 2019-03-04 16:11:55 2019-03-04 16:19:00           1      0.79   5.0  0.00   
2 2019-03-27 17:53:01 2019-03-27 18:00:25           1      1.37   7.5  2.36   
3 2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70  27.0  6.15   
4 2019-03-30 13:27:42 2019-03-30 13:37:14           3      2.16   9.0  1.10   

   tolls  total   color      payment            pickup_zone  \
0    0.0  12.95  yellow  credit card        Lenox Hill West   
1    0.0   9.30  yellow         cash  Upper West Side South   
2    0.0  14.16  yellow  credit card          Alphabet City   
3    0.0  36.95  yellow  credit card              Hudson Sq   
4    0.0  13.40  yellow  credit card           Midtown East   

            dropoff_zone pickup_borough dropoff_borough trip_duration_min  
0    UN/Turtle Bay South      Manhattan       Man

## **📌 Soal 2: Ubah Kolom Waktu & Buat Fitur Durasi**

Kolom `pickup` dan `dropoff` masih bertipe string.  
Ubahlah keduanya menjadi datetime, lalu buat kolom baru bernama `trip_duration_min` yang berisi durasi perjalanan dalam **menit**.

Setelah itu:
1. tampilkan 5 baris pertama dari kolom `pickup`, `dropoff`, dan `trip_duration_min`
2. tampilkan rata-rata durasi perjalanan


In [4]:
# Tulis jawabanmu di sini
df['pickup'] = pd.to_datetime(df['pickup'], errors='coerce')
df['dropoff'] = pd.to_datetime(df['dropoff'], errors='coerce')
df['trip_duration_min'] = df['dropoff'] - df['pickup']
avg_duration = df['trip_duration_min'].mean()
print("1. Hasil ubah ke date:\n", df[['pickup', 'dropoff']].dtypes)
print("\n2. 5 Baris pertama:\n", df.head())
print("\n3. Rata-rata durasi perjalanan:\n", avg_duration)

1. Hasil ubah ke date:
 pickup     datetime64[ns]
dropoff    datetime64[ns]
dtype: object

2. 5 Baris pertama:
                pickup             dropoff  passengers  distance  fare   tip  \
0 2019-03-23 20:21:09 2019-03-23 20:27:24           1      1.60   7.0  2.15   
1 2019-03-04 16:11:55 2019-03-04 16:19:00           1      0.79   5.0  0.00   
2 2019-03-27 17:53:01 2019-03-27 18:00:25           1      1.37   7.5  2.36   
3 2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70  27.0  6.15   
4 2019-03-30 13:27:42 2019-03-30 13:37:14           3      2.16   9.0  1.10   

   tolls  total   color      payment            pickup_zone  \
0    0.0  12.95  yellow  credit card        Lenox Hill West   
1    0.0   9.30  yellow         cash  Upper West Side South   
2    0.0  14.16  yellow  credit card          Alphabet City   
3    0.0  36.95  yellow  credit card              Hudson Sq   
4    0.0  13.40  yellow  credit card           Midtown East   

            dropoff_zone pickup_bo

## **📌 Soal 3: Seleksi Data dengan `loc`, `iloc`, dan Boolean Mask**

Kerjakan hal berikut:
1. Ambil kolom `fare`, `tip`, dan `total` untuk 10 baris pertama menggunakan `.loc`
2. Ambil 5 baris pertama dan 4 kolom pertama menggunakan `.iloc`
3. Tampilkan semua trip dengan:
   - `payment == "cash"`
   - `distance > 3`
   - `tip == 0`

Tampilkan 10 baris pertama dari hasil filter pada poin (3).


In [5]:
# Tulis jawabanmu di sini
data_satu = df[['fare', 'tip', 'total']].loc[:9]
data_dua = df.iloc[:5, :4]
bool_satu = df['payment'] == 'cash'
bool_dua = df['distance'] > 3
bool_tiga = df['tip'] == 0
bool_final = bool_satu & bool_dua & bool_tiga
print("1. Ambil kolom fare, tip, dan total untuk 10 baris pertama:\n", data_satu)
print("\n2. Ambil 5 baris pertama dan 4 kolom pertama:\n", data_dua)
print("\n3. Ambil semua baris dengan metode pembayaran cash, jarak lebih dari 3, dan tip 0:\n", df[bool_final].head(10))

1. Ambil kolom fare, tip, dan total untuk 10 baris pertama:
    fare   tip  total
0   7.0  2.15  12.95
1   5.0  0.00   9.30
2   7.5  2.36  14.16
3  27.0  6.15  36.95
4   9.0  1.10  13.40
5   7.5  2.16  12.96
6  13.0  2.00  18.80
7   8.5  0.00  11.80
8  15.0  1.00  19.30
9   8.0  1.00  13.30

2. Ambil 5 baris pertama dan 4 kolom pertama:
                pickup             dropoff  passengers  distance
0 2019-03-23 20:21:09 2019-03-23 20:27:24           1      1.60
1 2019-03-04 16:11:55 2019-03-04 16:19:00           1      0.79
2 2019-03-27 17:53:01 2019-03-27 18:00:25           1      1.37
3 2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70
4 2019-03-30 13:27:42 2019-03-30 13:37:14           3      2.16

3. Ambil semua baris dengan metode pembayaran cash, jarak lebih dari 3, dan tip 0:
                 pickup             dropoff  passengers  distance  fare  tip  \
10 2019-03-16 10:02:25 2019-03-16 10:22:29           1      3.90  17.0  0.0   
26 2019-03-03 02:45:36 2019-03-03

## **📌 Soal 4: Nilai Unik & Distribusi Kategori**

Tampilkan:
1. Semua nilai unik pada kolom `payment`
2. Jumlah nilai unik pada kolom `pickup_borough`
3. Banyaknya data pada setiap kategori `payment` menggunakan `value_counts()`

Lalu jawab dengan komentar: metode pembayaran apa yang paling sering muncul?


In [6]:
# Tulis jawabanmu di sini
print("1. Semua nilai unik pada kolom payment:\n", df['payment'].unique())
print("\n2. Jumlah nilai unik pada kolom pickup_borough:\n", df['pickup_borough'].nunique())
print("\n3. Banyaknya data pada setiap kategori payment menggunakan value_counts:\n", df['payment'].value_counts())

# Metode pembayaran paling sering muncul: Credit Card dengan total 4577

1. Semua nilai unik pada kolom payment:
 ['credit card' 'cash' nan]

2. Jumlah nilai unik pada kolom pickup_borough:
 4

3. Banyaknya data pada setiap kategori payment menggunakan value_counts:
 payment
credit card    4577
cash           1812
Name: count, dtype: int64


## **📌 Soal 5: Missing Value**

Kerjakan langkah berikut:
1. Tampilkan jumlah missing value per kolom
2. Buat salinan dataset bernama `df_clean`
3. Pada `df_clean`, isi missing value:
   - kolom `payment` dengan string `"unknown"`
   - kolom `dropoff_borough` dengan string `"Unknown"`
4. Tampilkan lagi jumlah missing value dari dua kolom tersebut setelah diisi

**Catatan:** gunakan `.copy()` supaya dataset asli `df` tetap aman.


In [7]:
# Tulis jawabanmu di sini
print("1. Tampilkan jumlah missing value per kolom:\n", df.isna().sum())
df_clean = df.copy()
print("\n2. Buat salinan dataset bernama df_clean:\n", df_clean)
df_clean['payment'].fillna('unknown')
df_clean['dropoff_borough'].fillna('Unknown')
print("\n3. Hasil isi missing value:\n", df_clean[(df_clean['payment'] == 'unknown') & (df_clean['dropoff_borough'] == 'Unknown')].head())
print("\n4. Jumlah missing value setelah diisi:\n", df_clean.isna().sum())


1. Tampilkan jumlah missing value per kolom:
 pickup                0
dropoff               0
passengers            0
distance              0
fare                  0
tip                   0
tolls                 0
total                 0
color                 0
payment              44
pickup_zone          26
dropoff_zone         45
pickup_borough       26
dropoff_borough      45
trip_duration_min     0
dtype: int64

2. Buat salinan dataset bernama df_clean:
                   pickup             dropoff  passengers  distance  fare  \
0    2019-03-23 20:21:09 2019-03-23 20:27:24           1      1.60   7.0   
1    2019-03-04 16:11:55 2019-03-04 16:19:00           1      0.79   5.0   
2    2019-03-27 17:53:01 2019-03-27 18:00:25           1      1.37   7.5   
3    2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70  27.0   
4    2019-03-30 13:27:42 2019-03-30 13:37:14           3      2.16   9.0   
...                  ...                 ...         ...       ...   ...   
6428 

## **📌 Soal 6: Data Duplikat**

Pada dataset `taxis`, data duplikat alami jumlahnya sedikit / bahkan bisa tidak ada.  
Supaya bisa latihan, buat dataset baru bernama `df_dupes` dengan cara menambahkan **15 baris acak dari `df_clean`** ke bagian bawah dataset.

Lalu:
1. Hitung jumlah baris duplikat pada `df_dupes`
2. Buang baris duplikat tersebut dan simpan ke `df_nodup`
3. Tampilkan shape sebelum dan sesudah menghapus duplikat


In [8]:
# Tulis jawabanmu di sini
df_dupes = pd.concat([df_clean, df_clean.sample(n=15, random_state=42)], ignore_index=True)
print("1. Jumlah baris duplikat pada df_dupes:\n", df_dupes.duplicated().sum())
df_nodupes = df_dupes.drop_duplicates().copy()
print("\n2. Buang baris ke df_nodup:\n", df_nodupes)
print("\n3. Jumlah baris sebelum duplikat dibuang:\n", df_dupes.shape[0],"\nsetelah duplikat dibuang:\n", df_nodupes.shape[0])

1. Jumlah baris duplikat pada df_dupes:
 15

2. Buang baris ke df_nodup:
                   pickup             dropoff  passengers  distance  fare  \
0    2019-03-23 20:21:09 2019-03-23 20:27:24           1      1.60   7.0   
1    2019-03-04 16:11:55 2019-03-04 16:19:00           1      0.79   5.0   
2    2019-03-27 17:53:01 2019-03-27 18:00:25           1      1.37   7.5   
3    2019-03-10 01:23:59 2019-03-10 01:49:51           1      7.70  27.0   
4    2019-03-30 13:27:42 2019-03-30 13:37:14           3      2.16   9.0   
...                  ...                 ...         ...       ...   ...   
6428 2019-03-31 09:51:53 2019-03-31 09:55:27           1      0.75   4.5   
6429 2019-03-31 17:38:00 2019-03-31 18:34:23           1     18.74  58.0   
6430 2019-03-23 22:55:18 2019-03-23 23:14:25           1      4.14  16.0   
6431 2019-03-04 10:09:25 2019-03-04 10:14:29           1      1.12   6.0   
6432 2019-03-13 19:31:22 2019-03-13 19:48:02           1      3.85  15.0   

       tip  t

## **📌 Soal 7: Feature Engineering Sederhana**

Gunakan `df_nodup`, lalu buat:
1. Kolom `tip_pct` dengan rumus:
 
$$\text{tip_pct} = \frac{tip}{fare} \times 100$$

2. Kolom `is_long_trip`:
   - `"yes"` jika `distance >= 5`
   - `"no"` jika `distance < 5`

3. Kolom `pickup_hour` dari waktu `pickup`

Tampilkan 10 baris pertama dari kolom:
`fare`, `tip`, `tip_pct`, `distance`, `is_long_trip`, `pickup_hour`


In [21]:
# Tulis jawabanmu di sini
df_nodupes['tip_pct'] = df_nodupes['tip'] / df_nodupes['fare'] * 100
def long_trip(duration):
    if duration >= 5:
        return 'yes'
    else:
        return 'no'
df_nodupes['is_long_trip'] = df_nodupes['distance'].apply(long_trip)
df_nodupes['pickup_hour'] = df_nodupes['pickup'].dt.hour
print("1. Kolom tip_pct:\n", df_nodupes['tip_pct'].head())
print("\n2. Kolom is_long_trip:\n", df_nodupes['is_long_trip'].head())
print("\n3. Kolom pickup_hour:\n", df_nodupes['pickup_hour'].head())
df_nodupes[['fare','tip','tip_pct','distance','is_long_trip','pickup_hour']].head(10)

1. Kolom tip_pct:
 0    30.714286
1     0.000000
2    31.466667
3    22.777778
4    12.222222
Name: tip_pct, dtype: float64

2. Kolom is_long_trip:
 0     no
1     no
2     no
3    yes
4     no
Name: is_long_trip, dtype: object

3. Kolom pickup_hour:
 0    20
1    16
2    17
3     1
4    13
Name: pickup_hour, dtype: int32


,fare,tip,tip_pct,distance,is_long_trip,pickup_hour
0,7.0,2.15,30.714286,1.60,no,20
1,5.0,0.00,0.000000,0.79,no,16
2,7.5,2.36,31.466667,1.37,no,17
3,27.0,6.15,22.777778,7.70,yes,1
4,9.0,1.10,12.222222,2.16,no,13
5,7.5,2.16,28.800000,0.49,no,10
6,13.0,2.00,15.384615,3.65,no,21
7,8.5,0.00,0.000000,1.40,no,12
8,15.0,1.00,6.666667,3.63,no,11
9,8.0,1.00,12.500000,1.52,no,16


## **📌 Soal 8: GroupBy & Aggregation**

Gunakan `df_nodup`, lalu buat ringkasan per `payment` yang berisi:
1. jumlah trip
2. rata-rata `fare`
3. rata-rata `tip`
4. rata-rata `trip_duration_min`

Urutkan hasilnya berdasarkan jumlah trip dari yang terbesar ke yang terkecil.


In [25]:
# Tulis jawabanmu di sini
jlh_trip = df_nodupes.groupby('payment')['distance'].size()
rata_fare = df_nodupes.groupby('payment')['fare'].mean()
rata_tip = df_nodupes.groupby('payment')['tip'].mean()
rata_trip_duration = df_nodupes.groupby('payment')['trip_duration_min'].mean()
order = jlh_trip.sort_values(ascending=False).index
print("1. Total trip per metode pembayaran:\n", jlh_trip.loc[order])
print("\n2. Rata-rata fare per metode pembayaran:\n", rata_fare.loc[order])
print("\n3. Rata-rata tip per metode pembayaran:\n", rata_tip.loc[order])
print("\n4. Rata-rata durasi perjalanan per metode pembayaran:\n", rata_trip_duration.loc[order])

1. Total trip per metode pembayaran:
 payment
credit card    4577
cash           1812
Name: distance, dtype: int64

2. Rata-rata fare per metode pembayaran:
 payment
credit card    13.694750
cash           11.592991
Name: fare, dtype: float64

3. Rata-rata tip per metode pembayaran:
 payment
credit card    2.781805
cash           0.000000
Name: tip, dtype: float64

4. Rata-rata durasi perjalanan per metode pembayaran:
 payment
credit card   0 days 00:15:04.060738474
cash          0 days 00:12:37.759381898
Name: trip_duration_min, dtype: timedelta64[ns]


## **📌 Soal 9: GroupBy + `transform()`**

Masih menggunakan `df_nodup`, buat kolom baru bernama `fare_vs_payment_mean` dengan rumus:

\[
fare\_vs\_payment\_mean = fare - \text{rata-rata fare pada grup payment yang sama}
\]

Artinya, setiap baris dibandingkan terhadap rata-rata `fare` dari metode pembayaran yang sama.

Tampilkan 10 baris pertama dari kolom:
`payment`, `fare`, `fare_vs_payment_mean`


In [11]:
# Tulis jawabanmu di sini
rata = df_nodupes.groupby('payment')['fare'].transform('mean')
df_nodupes['fare_vs_payment_mean'] = df_nodupes['fare'] / rata
df_nodupes[['payment', 'fare', 'fare_vs_payment_mean']].head(10)

,payment,fare,fare_vs_payment_mean
0,credit card,7.0,0.511145
1,cash,5.0,0.431295
2,credit card,7.5,0.547655
3,credit card,27.0,1.971558
4,credit card,9.0,0.657186
5,credit card,7.5,0.547655
6,credit card,13.0,0.949269
7,NaN,8.5,NaN
8,credit card,15.0,1.095310
9,credit card,8.0,0.584165


## **📌 Soal 10: Sorting & Menyimpan Hasil**

Buat dataset akhir bernama `final_df` dari `df_nodup` dengan langkah:
1. Urutkan berdasarkan `total` dari terbesar ke terkecil
2. Jika nilai `total` sama, urutkan `trip_duration_min` dari terkecil ke terbesar
3. Simpan hasilnya ke file bernama **`taxis_final.csv`**

Tampilkan 10 baris pertama dari `final_df`.


In [12]:
# Tulis jawabanmu di sini
final_df = df_nodupes.copy()
print("1. Hasil urutkan berdasarkan total dari terbesar:\n", final_df[['total']].sort_values(by='total', ascending=False).head(10))
print("\n2. Hasil double grouping:\n", final_df[['total','trip_duration_min']].sort_values(by=['total', 'trip_duration_min'], ascending=[False, True]).head(10))
final_df.to_csv('taxis_final.csv', index=False)

1. Hasil urutkan berdasarkan total dari terbesar:
        total
5364  174.82
5648  169.70
622   166.00
4050  144.30
2231  136.56
3644  123.95
4815  123.36
4219  113.56
3500  107.16
4218  100.49

2. Hasil double grouping:
        total trip_duration_min
5364  174.82   0 days 01:04:51
5648  169.70   0 days 01:29:53
622   166.00   0 days 00:00:19
4050  144.30   0 days 00:55:57
2231  136.56   0 days 00:46:24
3644  123.95   0 days 00:01:43
4815  123.36   0 days 00:25:22
4219  113.56   0 days 01:02:06
3500  107.16   0 days 01:09:45
4218  100.49   0 days 01:21:32


## **🌟 Bonus (Opsional): Analisis Singkat**

Pilih **satu** insight sederhana dari dataset ini, lalu tulis dalam markdown/cell komentar.  
Contoh:
- payment apa yang paling sering dipakai?
- borough mana yang paling sering jadi titik pickup?
- apakah trip yang lebih jauh selalu menghasilkan tip lebih besar?

Tidak perlu panjang, cukup 2–4 kalimat.


In [13]:
jlh_payment = df_nodupes.groupby('payment')['tip'].count()
print("Urutan metode pembayaran berdasarkan jumlah tip yang diberikan:\n", jlh_payment.sort_values(ascending=False))

Urutan metode pembayaran berdasarkan jumlah tip yang diberikan:
 payment
credit card    4577
cash           1812
Name: tip, dtype: int64


**Insight**

Dari hasil hipotesis di atas dimana saya mengelompokkan metode pembayaran atau 'payment' kemudian menghitung jumlah baris pada kolom 'tip', insight yang didapatkan adalah metode pembayaran yang paling sering digunakan adalah 'credit card' dengan jumlah pemakaian yaitu 4577. Metode pembayaran lainnya yang digunakan adalah 'cash' yang menempati top 2 metode pembayaran yang paling sering dipakai dengan total 1812 pemakaian dan juga terdapat 44 baris data yang tidak mencantumkan metode pembayaran yang digunakan dan direpresentasikan dengan 'unknown'

# **🎯 Penutup**

Kalau kamu sudah menyelesaikan semua soal:
- cek lagi hasil outputmu,
- pastikan bentuk array dan logikanya sesuai,
- lalu simpan notebook ini sebelum dikumpulkan.

Semangat! 🚀
